diff -aurp tgt-1.0.4/usr/iscsi/iscsid.c tgt-1.0.4.work/usr/iscsi/iscsid.c
--- tgt-1.0.4/usr/iscsi/iscsid.c	2011-01-17 20:09:28.000000000 -0600
+++ tgt-1.0.4.work/usr/iscsi/iscsid.c	2011-01-17 20:09:50.000000000 -0600
@@ -1943,6 +1943,13 @@ again:
 		conn->req.ahssize = conn->req.bhs.hlength * 4;
 		conn->req.datasize = ntoh24(conn->req.bhs.dlength);
 		conn->rx_size = conn->req.ahssize;
+
+		if (conn->state != STATE_SCSI &&
+		    conn->req.ahssize > INCOMING_BUFSIZE) {
+			conn->state = STATE_CLOSE;
+			return;
+		}
+
 		if (conn->rx_size) {
 			conn->rx_buffer = conn->req.ahs;
 			conn->rx_iostate = IOSTATE_RX_AHS;
@@ -1995,6 +2002,14 @@ again:
 		if (conn->rx_size) {
 			conn->rx_iostate = IOSTATE_RX_DATA;
 			conn->rx_buffer = conn->req.data;
+
+			if (conn->state != STATE_SCSI) {
+				if (conn->req.ahssize + conn->rx_size >
+				    INCOMING_BUFSIZE) {
+					conn->state = STATE_CLOSE;
+					return;
+				}
+			}
 		} else {
 			conn->rx_iostate = IOSTATE_RX_END;
 			break;
Only in tgt-1.0.4.work/usr/iscsi: iscsid.c.orig
