diff --git a/README b/README
index 22785fc..22b963f 100644
--- a/README
+++ b/README
@@ -9,17 +9,6 @@ Currently, tgt supports the following target drivers:
 
 - iSER software target driver for Infiniband and RDMA NICs
 
-- IBM System p VIO server
-
-- FCoE software target driver for Ethernet NICs (in progress)
-
-- Qlogic qla2xxx FC target driver (in progress)
-
-Tgt consists of kernel modules, user-space daemon, and user-space
-tools. iSCSI, iSER, and FCoE target drivers use only user-space daemon
-and tools (i.e. they are just user-space applications. They don't need
-any kernel support).
-
 tgt can emulate the following device types:
 
 - SBC: a virtual disk drive that can use a file to store the content.
diff --git a/doc/README.iscsi b/doc/README.iscsi
index 321b1d8..cd63de2 100644
--- a/doc/README.iscsi
+++ b/doc/README.iscsi
@@ -1,25 +1,27 @@
 Preface
 -------------
-This show a simple example to set up some targets.
+This shows a simple example to set up some targets.
 
+See man tgt-admin and the example /etc/tgt/targets.conf
+file for how to setup a persistent configuration that is
+started when the tgtd service is started (when "service tgtd start"
+is run).
 
 Starting the daemon
 -------------
 The iSCSI target driver works with the 2.6.X kernels.
 
-First, you need to compile the source code:
-
-host:~/tgt/usr$ make ISCSI=1
-
 Try the following commands:
 
 host:~/tgt$ su
-host:~/tgt# ./usr/tgtd
+host:~/tgt# service tgtd start
 
 
 Configuration
 -------------
-Everyting is configured via the tgtadm management tool.
+When started, the tgtd service will setup targets defined in
+/etc/tgt/targets.conf. To manually configure a target tgtadm
+or tgt-admin can be used.
 
 The following example creates a target with id 1 (the iqn is
 iqn.2001-04.com.example:storage.disk2.amiens.sys1.xyz) and adds a
diff --git a/doc/README.iser b/doc/README.iser
index 2361c4e..92edc61 100644
--- a/doc/README.iser
+++ b/doc/README.iser
@@ -1,6 +1,21 @@
 iSCSI Extensions for RDMA (iSER)
 ================================
 
+This is an detailed description of the iSER tgtd target. It
+covers issues from the design to how to manually set it up.
+
+NOTE:
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+To run this iSER target you must have installed the libiverbs
+and librdma rpms on your system. They will not get brought in
+automatically when installing this rpm.
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+See man tgt-admin and the example /etc/tgt/targets.conf
+file for how to setup a persistent configuration that is
+started when the tgtd service is started (when "service tgtd start"
+is run).
+
 Copyright (C) 2007 Pete Wyckoff <pw@osc.edu>
 
 Background
@@ -156,28 +171,6 @@ the Linux initiator does not add padding either.
 Using iSER
 ----------
 
-Compile tgtd with "make ISCSI=1 ISCSI_RDMA=1" to build iSCSI and iSER.
-You'll need to have two libraries installed on your system:
-libibverbs.so and librdmacm.so.  If they are installed in the normal
-system paths (/usr/include and /usr/lib or /usr/lib64), they will be
-found automatically.  Otherwise, edit CFLAGS and LIBS in usr/Makefile
-near ISCSI_RDMA to specify the paths by hand, e.g., for a /usr/local
-install, it should look like:
-
-	ifneq ($(ISCSI_RDMA),)
-	CFLAGS += -DISCSI_RDMA -I/usr/local/include
-	TGTD_OBJS += iscsi/iscsi_rdma.o
-	LIBS += -L/usr/local/lib -libverbs -lrdmacm
-	endif
-
-If these libraries are not in the normal system paths, you may
-possibly also have to set, e.g., LD_LIBRARY_PATH=/usr/local/lib
-in your environment to find the shared libraries at runtime.
-
-The target will listen on all TCP interfaces (as usual), as well as all
-RDMA devices.  Both use the same default iSCSI port, 3260.  Clients on
-TCP or RDMA will connect to the same tgtd instance.
-
 Start the daemon (as root):
 
 	./tgtd
